numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
     0,                                          /* tp_version_tag */
 };
 
+/* ______________________________________________________________________
+ * DUFunc: A call-time (hence dynamic) specializable ufunc.
+ */
+
+typedef struct {
+    PyObject_HEAD
+    PyObject      * dispatcher;
+    PyUFuncObject * ufunc;
+    PyObject      * keepalive;
+    int             frozen;
+} PyDUFuncObject;
+
+static void
+dufunc_dealloc(PyDUFuncObject *self)
+{
+    /* Note: There is no need to call PyArray_free() on
+       self->ufunc->ptr, since ufunc_dealloc() will do it for us. */
+    Py_XDECREF(self->ufunc);
+    Py_XDECREF(self->dispatcher);
+    Py_XDECREF(self->keepalive);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *
+dufunc_repr(PyDUFuncObject *dufunc)
+{
+#if PY_MAJOR_VERSION >= 3
+    return PyUnicode_FromFormat("<numba._DUFunc '%s'>", dufunc->ufunc->name);
+#else
+    return PyString_FromFormat("<numba._DUFunc '%s'>", dufunc->ufunc->name);
+#endif
+}
+
+static PyObject *
+dufunc_call(PyDUFuncObject *self, PyObject *args, PyObject *kws)
+{
+    PyObject *result=NULL;
+
+    result = PyUFunc_Type.tp_call((PyObject *)self->ufunc, args, kws);
+    if ((!self->frozen) &&
+            (result == NULL) &&
+            (PyErr_Occurred()) &&
+            (PyErr_ExceptionMatches(PyExc_TypeError))) {
+
+        /* Break back into Python when we fail at dispatch. */
+        PyErr_Clear();
+        PyObject *method = PyObject_GetAttrString(
+            (PyObject*)self, "_compile_for_args");
+
+        if (method) {
+            result = PyObject_Call(method, args, kws);
+            if (result) {
+                Py_DECREF(result);
+                result = PyUFunc_Type.tp_call((PyObject *)self->ufunc, args,
+                                              kws);
+            }
+        }
+        Py_XDECREF(method);
+    }
+    return result;
+}
+
+static Py_ssize_t
+_get_nin(PyObject * py_func_obj)
+{
+    int result = -1;
+    PyObject *inspect=NULL, *getargspec=NULL, *argspec=NULL, *args=NULL;
+
+    inspect = PyImport_ImportModule("inspect");
```

Comment:
Instead of having dedicated C code for this, why not pass the value at construction time?